native/src/main/java/io/ballerina/stdlib/xmldata/utils/XmlDataUtils.java

```diff
@@ -200,7 +200,11 @@ public static Type getTypeFromUnionType(Type childType, Object value) {
             for (Type memberType : bUnionType.getMemberTypes()) {
                 if (value.getClass().getName().toUpperCase(Locale.ROOT).contains(
                         memberType.getName().toUpperCase(Locale.ROOT))) {
-                    childType = memberType;
+                    if (memberType.getTag() == TypeTags.TYPE_REFERENCED_TYPE_TAG) {
+                        childType = TypeUtils.getReferredType(memberType);
+                    } else {
+                        childType = memberType;
+                    }
```

Comment:
```suggestion
                    childType = TypeUtils.getReferredType(memberType);
```
The `if` part is already handled by the `TypeUtils` API.